### PR TITLE
login, logout, and authentication over states implemented

### PR DESCRIPTION
**Pull request by osvaldoFernandez92**
Summary
Login and logout functionality. Using LocalStorageService to set/get token. Created AuthenticationService to log in / log out.

Test plan
- Without an account created, click on link "Edit user"
- Forbidden site alert should appear
- Create account / Login
- "Login and Signup links should dissapear in Home.index and Logout button should appear"
- When going to "Edit User"... "Hi" header title should appear.

### CODE REVIEW

**Comment by osvaldoFernandez92** [comment]
@sbalay

File: src/app/app.authorization.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+angular.module('app').run([
+  '$rootScope', '$state', 'authenticationService', '$http', 'localStorageService',
+  function ($rootScope, $state, authenticationService) {
+
+  $rootScope.$on('$stateChangeStart',
+    function (event, toState, toParams, fromState, fromParams) {
+      if (!authenticationService.isLoggedIn() && toState.data.requireLogin) {
```
Comment (sbalay) [inline src/app/app.authorization.js:7]: What if a logged user tries to go to the login page?

File: src/app/app.routes.js [hunk 2]
```diff
@@ -52,7 +55,50 @@ app.config([
...
+      abstract: true,
+      template: '<ui-view/>',
+      views: {
+        main: {
+          templateUrl: '../app/layouts/user/main.html'
+        }
+      }
+    })
+    .state('user.edit', {
+      url: '/user/edit',
```
Comment (sbalay) [inline src/app/app.routes.js]: set the 'user' url in the parent state

File: src/app/app.routes.js [hunk 3]
```diff
@@ -52,7 +55,50 @@ app.config([
...
+          templateUrl: '../app/components/login/login.html'
+        }
+      },
+      data: {
+        requireLogin: false
       }
+    }) 
+    .state('user', {
+      abstract: true,
+      template: '<ui-view/>',
```
Comment (sbalay) [inline src/app/app.routes.js]: this line is needless.

File: src/app/components/component1/home.jade [hunk 4]
```diff
@@ -1,5 +1,18 @@
...
+        a(ui-sref='login.index')
+          | Login
+      li
+        a(ui-sref='sign_up.index')
+          | Go to sing up
+  ul      
+    li
+      a(ui-sref='user.edit')
+        | Edit user
+  div(ng-show="homeCtrl.isLoggedIn()")
```
Comment (sbalay) [inline src/app/components/component1/home.jade:16]: needless validation

File: src/app/app.authorization.js (hunk 1 above)
Comment (osvaldoFernandez92) [inline src/app/app.authorization.js:7]: good question.. now I see why there's this validation in the authorization that you've written...

else if (authenticationService.isLoggedIn() && !toState.data.requireLogin)

however, in my training app, I have some states that doesnt require login and I dont want to "prevent default" in them if someone that is logged in enters.. one example of that is "home" state.. If someone goes to "home" I want them to see it, even if they are / are not logged in... How should I handle this? 
I thought in something like

else if (authenticationService.isLoggedIn() && !toState.url==='/login')
event.preventDefault();
alert('NOPE, you are already logged')
@sbalay , is my approach correct? :S

File: src/app/components/login/login.jade [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    .input-control
+      | Username
+      input(name='usrname', ng-model='loginCtrl.usrname', required)
+    .input-control
+      | Password
+      input(name='pw1', ng-model='loginCtrl.password', type='password', required, ng-minlength="6")
+    .input-control
+      input(type='submit' value='Submit')			
+
+  a(ui-sref='home.index')
```
Comment (sbalay) [inline src/app/components/login/login.jade:14]: Don't use a button. Redirect the user automatically in the success callback of the login

File: src/app/components/login/loginController.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+angular.module('app').controller('LoginController', [
+  'authenticationService',
+  function (authenticationService) {
+    
+    this.loginPhrase = 'Login';
+
+    const successCb = (data) => {
+      alert('You have logged in');
```
Comment (sbalay) [inline src/app/components/login/loginController.js:8]: Redirect the user to home

File: src/app/app.authorization.js (hunk 1 above)
Comment (sbalay) [inline src/app/app.authorization.js:7]: What about adding another flag like 'requireLogin' that will be used for these kind of states? The ones that can be accessed by either logged or not logged users.

File: src/app/components/component1/home.jade (hunk 4 above)
Comment (sbalay) [inline src/app/components/component1/home.jade:16]: If this view will be accessed by not logged users the validation is OK

File: src/app/components/login/login.jade (hunk 5 above)
Comment (sbalay) [inline src/app/components/login/login.jade:14]: Mmm, this is ok for now if the home will allow unlogged users.

**Comment by sbalay** [comment]
@osvaldoFernandez92

**Comment by osvaldoFernandez92** [comment]
@sbalay fixes applied

**Comment by sbalay** [comment]
:+1:

**Comment by sbalay** [comment]
@osvaldoFernandez92  close or merge this pull request.

**Pull request closed without merging**